certification/internal/shell/has_v2_related_image_schema_version.go

```diff
@@ -0,0 +1,230 @@
+package shell
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/operator-framework/api/pkg/manifests"
+	operatorsv1alpha1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
+	"github.com/redhat-openshift-ecosystem/openshift-preflight/certification"
+	"github.com/redhat-openshift-ecosystem/openshift-preflight/certification/errors"
+	"github.com/redhat-openshift-ecosystem/openshift-preflight/cli"
+	log "github.com/sirupsen/logrus"
+)
+
+// RelatedImagesAreSchemaVersion2Check is part of the Operator policy and implements
+// the Check interface.
+type RelatedImagesAreSchemaVersion2Check struct{}
+
+// Validate checks the image manifest for each related image referenced in a
+// ClusterServiceVersion and ensures that the schema version used is version 2.
+func (p *RelatedImagesAreSchemaVersion2Check) Validate(image string) (bool, error) {
+	imageToSchemaVersion, err := p.getDataToValidate(image)
+	if err != nil {
+		return false, fmt.Errorf("%w: %s", errors.ErrRunContainerFailed, err)
+	}
+
+	return p.validate(imageToSchemaVersion)
+}
+
+// getDataToValidate pulls a ClusterServiceVersion from the input operator bundle,
+// checks the ClusterServiceVersion's related images declaration, and assembles a list of
+// images and their image manifest schema version values.
+func (p *RelatedImagesAreSchemaVersion2Check) getDataToValidate(image string) (map[string]int, error) {
+	bundlePath, err := p.extractBundleFromImage(image)
+	if err != nil {
+		log.Error(err)
```

Comment:
Can we have an error message which describes what goes wrong?

Reply:
The `extractBundleFromImage` method logs stdout and stderr if it encounters any errors along the way (three total tasks happening in that method, each being checked for errors - although one is now deferred and only emits warnings).

The next line of this `if` block returns the error message within the context of the calling function which also logs. 

Actually, I feel like I might be double logging...! What do you think?

Reply:
I see. From a developer perspective, it might be good to see what is causing the issue. It might facilitate the troubleshooting.